adapter/internal/oasparser/model/async_api.go

```diff
@@ -0,0 +1,157 @@
+	Bindings        map[string]interface{} `json:"bindings,omitempty"`
+}
+
+// ChannelItem in AsyncAPI channels
+type ChannelItem struct {
+	Ref              string                 `json:"$ref,omitempty"`
+	Subscribe        interface{}            `json:"subscribe,omitempty"` // TODO: (suksw) OperationAsync or $ref
+	Publish          interface{}            `json:"publish,omitempty"`   // TODO: (suksw) OperationAsync or $ref
+	Parameters       map[string]interface{} `json:"parameters,omitempty"`
+	Bindings         map[string]interface{} `json:"bindings,omitempty"`
+	VendorExtensions map[string]interface{} `json:"-"`
+}
+
+// OperationAsync is the Operation object that includes the message object
+type OperationAsync struct {
+	OperationID string `json:"operationId,omitempty"`
+	Message     struct {
+		Headers       string `json:"headers,omitempty"`
+		Payload       string `json:"payload,omitempty"`
+		CorrelationID string `json:"correlationId,omitempty"`
+		SchemaFormat  string `json:"schemaFormat,omitempty"`
+		ContentType   string `json:"contentType,omitempty"`
+		Name          string `json:"name,omitempty"`
+		Title         string `json:"title,omitempty"`
+	}
+}
+
+// SetInfoAsyncAPI populates the MgwSwagger object with information in asyncapi.yaml.
+func (swagger *MgwSwagger) SetInfoAsyncAPI(asyncAPI AsyncAPI) error {
+	swagger.vendorExtensions = asyncAPI.VendorExtensions
+	swagger.securityScheme = asyncAPI.getSecuritySchemes()
+	swagger.resources = asyncAPI.getResources()
+
+	if asyncAPI.Servers.Production.URL != "" {
+		endpoint, err := getWebSocketEndpoint(asyncAPI.Servers.Production.URL)
+		if err == nil {
+			productionEndpoints := append([]Endpoint{}, *endpoint)
+			swagger.productionEndpoints = generateEndpointCluster(constants.ProdClustersConfigNamePrefix,
+				productionEndpoints, constants.LoadBalance)
+		} else {
+			return errors.New("error encountered when parsing the production endpoint for AsyncAPI")
+		}
+	}
+	if asyncAPI.Servers.Sandbox.URL != "" {
+		endpoint, err := getWebSocketEndpoint(asyncAPI.Servers.Sandbox.URL)
+		if err == nil {
+			sandboxEndpoints := append([]Endpoint{}, *endpoint)
+			swagger.sandboxEndpoints = generateEndpointCluster(constants.SandClustersConfigNamePrefix,
+				sandboxEndpoints, constants.LoadBalance)
+		} else {
+			return errors.New("error encountered when parsing the sandbox endpoint for AsyncAPI")
+		}
+	}
+	return nil
+}
+
+func (asyncAPI AsyncAPI) getSecuritySchemes() []SecurityScheme {
+	securitySchemes := []SecurityScheme{}
+	for key, securityScheme := range asyncAPI.Components.SecuritySchemes {
+		securityScheme.DefinitionName = key
+		securitySchemes = append(securitySchemes, securityScheme)
+	}
+	return securitySchemes
+}
+
+func (asyncAPI AsyncAPI) getResources() []*Resource {
+	resources := []*Resource{}
+	for channel, channelItem := range asyncAPI.Channels {
+		var methodsArray []*Operation
+		methodFound := false
+		if channelItem.Publish != nil {
```

Comment:
In the envoy-context, publish and subscribe both methods are treated in the same manner. But we do not need to duplicate the same resource if both publish and subscribe is listed. 
In addition, we need to prioritize one over the other, because there could be occasions where some extensions mentioned under one operation but not in other.

Reply:
Prioritizing "Publish" when both types are defined. Fixed with https://github.com/wso2/product-microgateway/pull/2648/commits/a62d44caa9cbfad105440a12fe2851a3d466efa2